cmd/server.go

```diff
@@ -103,31 +94,99 @@ func serverOptsFor(cmd *cobra.Command) (map[string]plugin.Root, server.Opts, err
 		return nil, server.Opts{}, err
 	}
 
-	// Unmarshal the external plugins, if any are specified
-	var externalPlugins []plugin.ExternalPluginSpec
-	if err := viper.UnmarshalKey("external-plugins", &externalPlugins); err != nil {
-		return nil, server.Opts{}, fmt.Errorf("failed to unmarshal the external-plugins key: %v", err)
-	}
-
-	// Load internal plugins that are not specifically excluded.
-	enabledPlugins := viper.GetStringSlice("plugins")
 	plugins := make(map[string]plugin.Root)
-	if len(enabledPlugins) > 0 {
+
+	// Check the internal plugins
+	if viper.IsSet("plugins") || viper.IsSet("external-plugins") {
+		enabledPlugins := viper.GetStringSlice("plugins")
 		for _, name := range enabledPlugins {
-			if plug, ok := internalPlugins[name]; ok {
+			if plug, ok := server.InternalPlugins[name]; ok {
 				plugins[name] = plug
 			} else {
 				log.Warnf("Requested unknown plugin %s", name)
 			}
 		}
-	} else {
-		// Copy internalPlugins so we don't mutate it.
-		for name, plug := range internalPlugins {
+	} else if !plugin.IsInteractive() {
+		// This is an edge-case for a user but a common case for
+		// CI. Thus, load all the plugins so that we don't break
+		// the latter. Note that we copy server.InternalPlugins
+		// so that we don't mutate it.
+		log.Warnf("Running non-interactively without having set the 'plugins'/'external-plugins' keys in %v. Loading all core plugins by default", configFile)
+		for name, plug := range server.InternalPlugins {
 			plugins[name] = plug
 		}
+	} else {
+		// Assume first-time user. First, we prompt them to get a list
+		// of plugins that they wish to enable. Next, we write the
+		// enabled plugins back to their specified config file.
+		enabledPlugins := []string{}
+
+		// Prompt them for the list of enabled plugins. This should look something
```

Comment:
Each of these steps would probably be nice as a separate function we could try to mock and test.

Reply:
Sure, I can break it out.